lib/Sema/TypeCheckAttr.cpp

```diff
@@ -4007,7 +4007,9 @@ static bool checkFunctionSignature(
   if (!std::equal(required->getParams().begin(), required->getParams().end(),
                   candidateFnTy->getParams().begin(),
                   [&](AnyFunctionType::Param x, AnyFunctionType::Param y) {
-                    return x.getOldType()->isEqual(mapType(y.getOldType()));
+                    auto xInstanceTy = x.getOldType()->getMetatypeInstanceType();
+                    auto yInstanceTy = y.getOldType()->getMetatypeInstanceType();
```

Comment:
when either `required` or `candidateFnTy` are defined as `static`, the parameters are MetaTypes, hence I try to look into the instance type to be able to make this comparison work. Otherwise, this will fail and trigger the diagnostic we see today. Do you see any issues with this here?

Reply:
I recommend checking for static-ness mismatches in differentiability witnesses before checking for signature mismatches. That'll let you diagnose type mismatches as a separate thing.

Reply:
The signature mismatches are done during the process of looking up for the original function that the derivative is trying to derivate, that's why I modified this check so that the lookup does discards the candidate due to static-ness mismatch. Once the original candidate is found, I can do the static mismatch checks I introduced in this PR.